### DISCUSSION TITLE
Has anyone successfully built an MSI using GitLab CI/CD and WIX 4?

### DISCUSSION THREAD

**Discussion by jforward5** [wixtoolset/issues#7979] [Questions]
I am getting close to a solution for my problem of using GitLab to build my MSI using WiX 4.  The problem is that when building my project locally everything works perfectly.  I've attached a snippet from my WiX project file that I believe to be the issue when executed in the CI/CD pipeline.

[wix-snippet.txt](https://github.com/wixtoolset/issues/files/14114185/wix-snippet.txt)

When this builds in the GitLab CI/CD i get the following error:
MSBuild version 17.7.4+3ebbd7c49 for .NET
MSBUILD : error MSB1009: Project file does not exist.

Originally I was getting errors about WiX not being found etc, but i was able to resolve those issues by adding the following config to my "build" step in the yaml file:
`build:`
`  image: mcr.microsoft.com/dotnet/sdk:7.0`
`  stage: build`
`  before_script:`
`  - dotnet tool install --global wix`
`  - export PATH="$PATH:/root/.dotnet/tools"`
`  - dotnet build --runtime win-x64 restore --packages $NUGET_PACKAGES_DIRECTORY`
`  script:`
`  - dotnet build --runtime win-x64 --no-restore`

The above config at least attempts the build.  Any ideas on how to get the MSI to build using GitLab CI/CD?

**Comment by barnson** [comment]
WiX doesn't run as-is on Linux; it needs Windows components outside .NET. You _might_ have better luck with a Windows container.

**Reply by jforward5** [reply]
From my understanding the containers I am using are windows based.
image: mcr.microsoft.com/dotnet/sdk:7.0

Looked them up on MS site and they show to be windows server based....

**Reply by barnson** [reply]
`export PATH="$PATH:/root/.dotnet/tools"` smells more like Linux than Windows.

**Reply by jforward5** [reply]
I'll do some more looking into the containers then.

**Reply by jforward5** [reply]
Ok so I just ran my pipeline on a dedicated windows container, and now I'm getting a new error.  But I'm not sure if this is a Pipeline error...

Using docker image sha256:943cd53982aec26319ff98771f81ee5fb2c1bcbbfe97998b0686abcbb28b92fb for mcr.microsoft.com/dotnet/sdk:7.0 with digest mcr.microsoft.com/dotnet/sdk@sha256:6b9c857a60b67f968fd5bae373e414690947b8b58e7951b3228fd712b82938e8 ...

$ dotnet tool install --global wix
Tools directory 'C:\Users\ContainerUser\.dotnet\tools' is not currently on the PATH environment variable.
You can add the directory to the PATH by running the following command:
setx PATH "%PATH%;C:\Users\ContainerUser\.dotnet\tools"
You can invoke the tool using the following command: wix
Tool 'wix' (version '4.0.3') was successfully installed.
$ setx PATH "%PATH%;C:\Users\ContainerUser\.dotnet\tools"
ERROR: The system cannot find registry path specified.

I also have a ticket with GitLab but they haven't responded yet...

**Comment by jforward5** [comment]
A little more info.  Found the correct "Windows" runner to use, but still getting the original error:

You can invoke the tool using the following command: wix
Tool 'wix' (version '4.0.3') was successfully installed.
$ setx PATH "C:\Users\ContainerUser\.dotnet\tools;%PATH%"
SUCCESS: Specified value was saved.
$ dotnet build --runtime win-x64 restore --packages $NUGET_PACKAGES_DIRECTORY
MSBuild version 17.8.3+195e7f5a3 for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: restore

**Reply by rsking** [reply]
That dotnet command makes no sense. You are telling it to build a project in a folder called 'restore'.

I presume that you are trying to restore the nuget packages. That would be done with the command

`dotnet restore --runtime win-x64 --packages $NUGET_PACKAGES_DIRECTORY`

**Reply by rsking** [reply]
I also do not think that you need to have the `wix` global tool to do this. The build should work with a `restore`, then `build`.

**Reply by barnson** [reply]
And `dotnet` does an implicit `restore` before `build`.

**Reply by jforward5** [reply]
I am now having a error related to short file names: 
`The short file name 'avv63avc.exe' was generated for multiple files that may be installed to the same directory. This could be due to conflicting long file names specified by the File/@Name attribute. If that is the case, please resolve the conflict in those attributes. Otherwise, please manually set the File/@ShortName attribute on the conflicting row to fix the collision. If one of the colliding files was added via a patch, that short file name should be specified manually to avoid disturbing the original short file name.`

How do I set the short name for a file, and how do I see which file or files is prompting this message?

**Reply by jforward5** [reply]
Never mind I figured out how to do the short name, but it seems to not be the issue.  The issue seems to be 2 files named the same getting placed into the folder.  I need to figure out how to exclude one of them.

**Comment by jforward5** [comment]
Thank you both!  I was able to get it to run finally in GitLab!  Now I need to find out why it is producing errors that don't exist when built on my machine.
